File: src/android/security/src/main/java/org/meshchat/storage/EncryptedStorage.kt
package org.meshchat.storage

import android.content.Context
import android.database.Cursor
import android.system.Os
import android.system.OsConstants
import androidx.sqlite.db.SimpleSQLiteQuery
import java.io.File
import net.zetetic.database.sqlcipher.SQLiteDatabase
import org.meshchat.identity.*
import uniffi.meshchat_core.*

/** A connection never escapes an unlocked operation; no exception text is logged. */
internal class CipherConnection(private val database: SQLiteDatabase) : SqlDatabase, AutoCloseable {
    private fun arguments(values: List<SqlValue>): Array<Any> = values.map {
        when(it) { is SqlValue.Integer -> it.value; is SqlValue.Bytes -> it.value; is SqlValue.Text -> it.value }
    }.toTypedArray()
    override fun execute(sql: String, values: List<SqlValue>) {
        try { database.execSQL(sql, arguments(values)) }
        catch (_: Exception) { throw StorageException.Database() }
    }
    override fun query(sql: String, values: List<SqlValue>, limit: UInt): List<SqlRow> {
        if (limit > 100u) throw StorageException.InvalidInput()
        try {
            database.query(SimpleSQLiteQuery(sql, arguments(values))).use { cursor ->
                val rows = mutableListOf<SqlRow>()
                while(cursor.moveToNext()) {
                    if(rows.size >= limit.toInt() || cursor.columnCount > 12) throw StorageException.Database()
                    val cells = (0 until cursor.columnCount).map { column -> when(cursor.getType(column)) {
                        Cursor.FIELD_TYPE_INTEGER -> SqlValue.Integer(cursor.getLong(column))
                        Cursor.FIELD_TYPE_BLOB -> SqlValue.Bytes(cursor.getBlob(column).also { if(it.size>4096) throw StorageException.Database() })
                        Cursor.FIELD_TYPE_STRING -> SqlValue.Text(cursor.getString(column).also { if(it.length>4096) throw StorageException.Database() })
                        else -> throw StorageException.Database()
                    } }
                    rows.add(SqlRow(cells))
                }
                return rows
            }
        } catch (_: Exception) { throw StorageException.Database() }
    }
    override fun close() { database.close() }
    companion object {
        fun open(file: File, key: ByteArray, create: Boolean): CipherConnection {
            System.loadLibrary("sqlcipher")
            val flags = SQLiteDatabase.OPEN_READWRITE or (if(create) SQLiteDatabase.CREATE_IF_NECESSARY else 0)
            val db = try { SQLiteDatabase.openDatabase(file.path, key, null, flags, null, null) }
                catch (_: Exception) { throw StorageException.Database() }
            try {
                val result = CipherConnection(db)
                result.query("PRAGMA journal_mode=DELETE", emptyList(), 1u)
                result.execute("PRAGMA synchronous=FULL", emptyList())
                return result
            } catch(e: Exception) { db.close(); throw e }
        }
    }
}

/** Construct before IdentityProvider and pass it as the mandatory reset callback.
 * All operations share the identity lock; no DB/key handle survives an operation. */
class StorageVault internal constructor(
    private val folder: File,
    private val files: IdentityStorage,
    private val protection: IdentityProtection,
) : IdentityResetStore {
    companion object {
        fun android(context: Context): StorageVault {
            val app=context.applicationContext
            return StorageVault(File(app.noBackupFilesDir,"meshchat-storage-v1"), AndroidIdentityStorage(app,"meshchat-storage-v1"),
                AndroidIdentityProtection(app,"org.meshchat.storage.wrapping.v1"))
        }
    }
    private fun sync(directory:File) {val fd=Os.open(directory.path,OsConstants.O_RDONLY,0);try{Os.fsync(fd)}finally{Os.close(fd)}}
    private fun refuse(failure:IdentityFailure):Nothing = throw IdentityProviderException(failure)
    private val database get()=File(folder,"history.db")
    internal fun <T> access(generation:ByteArray,create:Boolean,now:Long?,work:(EncryptedStore)->T):T = IdentityProvider.withOperation {
        protection.requireUnlocked()
        if(generation.size!=16) refuse(IdentityFailure.INVALID_INPUT)
        if(create) {
            if(files.hasArtifacts()||protection.exists()) refuse(IdentityFailure.RECOVERY_REQUIRED)
            files.write(IdentityFile.JOURNAL,byteArrayOf(1))
            protection.create()
            val key=protection.random(64)
            try {val header=byteArrayOf(1)+generation; files.write(IdentityFile.ENVELOPE,header+protection.seal(key,header))}
            finally{key.fill(0)}
        } else {
            if(files.exists(IdentityFile.JOURNAL)) refuse(IdentityFailure.RECOVERY_REQUIRED)
            if(!files.exists(IdentityFile.ENVELOPE)||!database.isFile) refuse(if(files.hasArtifacts()||protection.exists()) IdentityFailure.RECOVERY_REQUIRED else IdentityFailure.MISSING)
        }
        if(!protection.exists()) refuse(IdentityFailure.INVALIDATED)
        val encoded=files.read(IdentityFile.ENVELOPE)
        if(encoded.size !in 18..1024||encoded[0]!=1.toByte()||!encoded.copyOfRange(1,17).contentEquals(generation)) refuse(IdentityFailure.INVALID_INPUT)
        val key=protection.open(encoded.copyOfRange(17,encoded.size),encoded.copyOfRange(0,17))
        try {
            if(key.size!=64) refuse(IdentityFailure.INVALID_INPUT)
            val result=CipherConnection.open(database,key,create).use { connection ->
                val core=EncryptedStore.open(connection,generation,create,now)
                try { work(core) } finally {core.close()}
            }
            protection.requireUnlocked()
            if(create){sync(folder);files.delete(IdentityFile.JOURNAL)}
            result
        } finally{key.fill(0)}
    }
    override fun clearIdentityState() = IdentityProvider.withOperation {
        IdentityProvider.requireResetInProgress()
        protection.requireUnlocked()
        if(files.hasArtifacts()||protection.exists()) {
            files.write(IdentityFile.JOURNAL,byteArrayOf(2))
            protection.delete()
            for(name in listOf("history.db","history.db-wal","history.db-shm","history.db-journal")) {
                val file=File(folder,name);if(file.exists()&&!file.delete()) refuse(IdentityFailure.RECOVERY_REQUIRED)
            }
            files.delete(IdentityFile.ENVELOPE);files.delete(IdentityFile.JOURNAL)
            if(folder.exists()&&!folder.delete()) refuse(IdentityFailure.RECOVERY_REQUIRED)
            sync(checkNotNull(folder.parentFile))
        }
    }
}

class EncryptedStorage(private val identity:IdentityProvider,private val vault:StorageVault,private val staff:StaffKeyVault?=null) {
    private fun <T> operation(create:Boolean=false,now:Long?=System.currentTimeMillis()/1000,work:(EncryptedStore)->T):T = IdentityProvider.withOperation {
        val generation=identity.load().identity.generation
        vault.access(generation,create,now,work)
    }
    fun create() {operation(create=true){}}
    fun reopen() {operation{}}
    fun putRecord(kind:RecordKind,key:ByteArray,value:ByteArray) {operation{it.putRecord(kind,key,value)}}
    fun getRecord(kind:RecordKind,key:ByteArray):ByteArray? = operation{it.getRecord(kind,key)}
    fun deleteRecord(kind:RecordKind,key:ByteArray) {operation{it.deleteRecord(kind,key)}}
    fun appendUnverified(item:HistoryItem) {operation{it.appendUnverified(item,System.currentTimeMillis()/1000)}}
    fun acceptAuthenticated(item:HistoryItem,subject:ByteArray,immutableBytes:ByteArray):AcceptResult = operation{it.acceptAuthenticated(item,subject,immutableBytes,System.currentTimeMillis()/1000)}
    fun history(conversation:ByteArray,direct:Boolean,limit:UInt):List<HistoryItem> = operation{it.history(conversation,direct,limit)}
    fun deleteHistory(conversation:ByteArray,direct:Boolean) {operation{it.deleteHistory(conversation,direct)}}
    fun ambiguousTarget(subject:ByteArray,messageId:ByteArray):Boolean = operation{it.ambiguousTarget(subject,messageId)}
    fun prune(){operation{it.prune(System.currentTimeMillis()/1000)}}
    /** Only public identity/friend state escapes construction; the database and
     * wrapping key are closed/cleared before returning to the feature owner. */
    fun transport(instance: ULong, now: ULong): NativeTransport = operation { store ->
        NativeTransport(store, identity.load().identity, instance, now)
    }
    fun channelAccept(owner: NativeChannels, link: LinkHandle?, bytes: ByteArray, intake: TransportIntake,
        own: Boolean, now: ULong): Boolean = operation { store ->
        owner.accept(store, ChannelReceipt(link, bytes, intake, own, System.currentTimeMillis()/1000, now))
    }
    fun channelHistory(owner: NativeChannels, name: String, nickname: String): List<ChannelMessage> =
        operation { owner.history(it, name, nickname) }

    fun friendCards(core: NativeTransport, now: ULong): List<FriendCard> = operation { core.friendCards(it, now) }
    fun friendCode(nickname: String): FriendProposal = friendCode(identity.load().identity, nickname)
    fun confirmFriend(core: NativeTransport, uri: String, petname: String, previous: FriendHandle?, now: ULong): TransportEffects =
        operation { store -> identity.messaging { core.confirmFriend(store, it, uri, petname, previous, now) } }
    fun changeFriend(core: NativeTransport, friend: FriendHandle, replace: Boolean, now: ULong): TransportEffects =
        operation { core.changeFriend(it, friend, replace, now) }
    fun authenticate(core: NativeTransport, link: LinkHandle, bytes: ByteArray, now: ULong): AuthenticationResult =
        operation { store -> identity.messaging { core.authenticateMessage(store, it, link, bytes, now, System.currentTimeMillis()/1000) } }
    fun retryMessages(core: NativeTransport, now: ULong): Boolean =
        operation { store -> identity.messaging { core.retryMessages(store, it, now, System.currentTimeMillis()/1000) } }
    fun sendDirect(core: NativeTransport, friend: FriendHandle, content: DirectContent, cookie: ULong, now: ULong): MessageSubmission =
        operation { store -> identity.messaging { core.sendDirect(store, it, friend, content, cookie, now, System.currentTimeMillis()/1000) } }
    fun sendSigned(core: NativeTransport, bytes: ByteArray, cookie: ULong, now: ULong): MessageSubmission =
        operation { store -> identity.messaging { core.sendSigned(store, it, bytes, cookie, now, System.currentTimeMillis()/1000) } }
    fun directHistory(core: NativeTransport, keys: ByteArray, now: ULong): List<DirectMessage> =
        operation { core.directHistory(it, keys, now, System.currentTimeMillis()/1000) }
    fun messagingHistory(core: NativeTransport, owner: NativeChannels, name: String, nickname: String): List<ChannelMessage> =
        operation { core.messagingChannelHistory(it, owner, name, nickname) }
    fun proof(core: NativeTransport, link: LinkHandle, now: ULong): TransportEffects =
        IdentityProvider.withOperation { identity.messaging { core.prepareProof(link, it, now) } }
    /** Keep reset serialization through the actual native submission. The
     * database/session are closed before returning; callbacks never hold keys. */
    fun messageEgress(core: NativeTransport, send: TransportSend, submit: () -> Boolean): Boolean = IdentityProvider.withOperation {
        if (core.messageNeedsAuthorization(send.link, send.token)) {
            val organizer=operation {
                val guarded=core.authorizeOrganizerEgress(it,send.link,send.token,System.currentTimeMillis()/1000)
                if(!guarded)core.authorizeMessageEgress(it, send.link, send.token)
                guarded
            }
            if(organizer)return@withOperation staff?.submitForeground(submit) ?: false
        }
        submit()
    }

    fun events(core: NativeTransport): List<EventCard> = operation { core.eventCards(it,System.currentTimeMillis()/1000) }
    fun adoptEvent(core: NativeTransport,uri:String,now:ULong) = operation { core.confirmEvent(it,uri,now,System.currentTimeMillis()/1000) }
    fun removeEvent(core:NativeTransport,key:ByteArray) = operation { core.removeEvent(it,key,System.currentTimeMillis()/1000) }
    fun eventMessages(core:NativeTransport,now:ULong):List<EventMessage> = operation { core.eventMessages(it,System.currentTimeMillis()/1000,now) }
    fun importStaff(core:NativeTransport,candidate:ByteArray,now:ULong) = operation { store ->
        checkNotNull(staff).importConfirmed(identity.load().identity.generation,candidate) { uri ->
            core.importStaffKey(store,uri,now,System.currentTimeMillis()/1000).use { it.invalidate() }
        }
    }
    fun staffCard():StaffCard? {
        val owner=staff ?: return null
        if(!owner.present())return null
        return IdentityProvider.withOperation { owner.access(identity.load().identity.generation,::staffProposal) }
    }
    fun forgetStaff(core:NativeTransport) = IdentityProvider.withOperation {
        core.forgetStaffOperations();checkNotNull(staff).forget()
    }
    fun organizerTick(core:NativeTransport,now:ULong):TransportEffects = operation {
        val wall=System.currentTimeMillis()/1000
        val card=try {if(staff?.signingAllowed()==true)staffCard() else null} catch (_:IdentityProviderException) {null}
        core.organizerTick(it,card?.takeIf {c->wall in c.notBefore.toLong()..c.notAfter.toLong()}?.credential,now,wall)
    }
    fun postEvent(core:NativeTransport,nickname:String,text:String,avatar:UByte,pin:UInt?,cookie:ULong,now:ULong):MessageSubmission = operation { store ->
        checkNotNull(staff).access(identity.load().identity.generation) { uri ->
            core.importStaffKey(store,uri,now,System.currentTimeMillis()/1000).use { session ->
                try { core.postEvent(store,session,nickname,text,avatar,pin,cookie,now,System.currentTimeMillis()/1000) }
                finally {session.invalidate()}
            }
        }
    }

    /** Bounded encrypted navigation index for retained old-identity histories.
     * Persist before pin mutation; a failed mutation can leave a harmless duplicate.
     * Eight slots hold eight canonical public-key codes each, below 2 KiB/record. */
    fun archives(): List<FriendProposal> = operation { store -> (0 until 8).flatMap { slot ->
        store.getRecord(RecordKind.SETTING, "ui-old-friends-$slot".toByteArray())?.toString(Charsets.UTF_8)
            ?.split('\n')?.filter { it.isNotEmpty() }?.map(::friendProposal) ?: emptyList()
    } }
    fun archiveFriend(friend: FriendCard): Boolean = operation { store ->
        val slots=(0 until 8).map { slot ->
            store.getRecord(RecordKind.SETTING, "ui-old-friends-$slot".toByteArray())?.toString(Charsets.UTF_8)
                ?.split('\n')?.filter { it.isNotEmpty() }?.toMutableList() ?: mutableListOf()
        }
        if (slots.flatten().any { friendProposal(it).keys.contentEquals(friend.keys) }) true else {
            val slot=slots.indexOfFirst { it.size<8 }
            if(slot<0) false else {
                val metadata=PublicIdentity(byteArrayOf(),friend.keys.copyOfRange(0,32),friend.keys.copyOfRange(32,64),byteArrayOf(),byteArrayOf())
                slots[slot].add(friendCode(metadata,friend.petname).uri)
                store.putRecord(RecordKind.SETTING,"ui-old-friends-$slot".toByteArray(),slots[slot].joinToString("\n").toByteArray());true
            }
        }
    }
    fun deleteArchive(keys: ByteArray) = operation { store ->
        store.deleteHistory(keys,true)
        for(slot in 0 until 8) {
            val name="ui-old-friends-$slot".toByteArray()
            val old=store.getRecord(RecordKind.SETTING,name)?.toString(Charsets.UTF_8) ?: continue
            val kept=old.split('\n').filter { it.isNotEmpty() && !friendProposal(it).keys.contentEquals(keys) }
            store.putRecord(RecordKind.SETTING,name,kept.joinToString("\n").toByteArray())
        }
    }
}

File: src/ios/Security/EncryptedStorage.swift
import Foundation
import Darwin
import SQLCipher

// The connection is private to a synchronous, unlocked native operation.
// A lock covers callback entry; callers receive typed store operations only.
final class CipherConnection: SqlDatabase, @unchecked Sendable {
    private var db: OpaquePointer?
    private let lock = NSRecursiveLock()
    init(file: URL, key: Data, create: Bool) throws {
        var handle: OpaquePointer?
        let rc = sqlite3_open_v2(file.path, &handle, SQLITE_OPEN_READWRITE | (create ? SQLITE_OPEN_CREATE : 0) | SQLITE_OPEN_FULLMUTEX, nil)
        guard rc == SQLITE_OK, let handle else { if let handle { sqlite3_close(handle) }; throw StorageError.Database }
        db = handle
        do {
            let keyed = key.withUnsafeBytes { sqlite3_key(handle, $0.baseAddress, Int32($0.count)) }
            guard keyed == SQLITE_OK else { throw StorageError.Database }
            _ = try query(sql: "PRAGMA journal_mode=DELETE", values: [], limit: 1)
            try execute(sql: "PRAGMA synchronous=FULL", values: [])
        } catch { close(); throw error }
    }
    func close() { lock.lock(); defer { lock.unlock() }; if let db { sqlite3_close_v2(db) }; db = nil }
    deinit { if let db { sqlite3_close_v2(db) } }
    private func prepare(_ sql: String, _ values: [SqlValue]) throws -> OpaquePointer {
        guard let db else { throw StorageError.Unavailable }
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK, let statement else {
            if let statement { sqlite3_finalize(statement) }; throw StorageError.Database
        }
        do {
            guard sqlite3_bind_parameter_count(statement) == values.count else { throw StorageError.InvalidInput }
            let transient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
            for (offset, value) in values.enumerated() {
                let index = Int32(offset + 1); let result: Int32
                switch value {
                case .integer(let value): result = sqlite3_bind_int64(statement, index, value)
                case .text(let value): result = value.withCString { sqlite3_bind_text(statement, index, $0, -1, transient) }
                case .bytes(let value):
                    result = value.isEmpty ? sqlite3_bind_zeroblob(statement, index, 0) : value.withUnsafeBytes { sqlite3_bind_blob(statement, index, $0.baseAddress, Int32($0.count), transient) }
                }
                guard result == SQLITE_OK else { throw StorageError.Database }
            }
            return statement
        } catch { sqlite3_finalize(statement); throw error }
    }
    func execute(sql: String, values: [SqlValue]) throws {
        lock.lock(); defer { lock.unlock() }
        let statement = try prepare(sql, values); defer { sqlite3_finalize(statement) }
        guard sqlite3_step(statement) == SQLITE_DONE else { throw StorageError.Database }
    }
    func query(sql: String, values: [SqlValue], limit: UInt32) throws -> [SqlRow] {
        lock.lock(); defer { lock.unlock() }
        guard limit <= 100 else { throw StorageError.InvalidInput }
        let statement = try prepare(sql, values); defer { sqlite3_finalize(statement) }
        var rows: [SqlRow] = []
        while true {
            let rc = sqlite3_step(statement)
            if rc == SQLITE_DONE { return rows }
            guard rc == SQLITE_ROW, rows.count < limit, sqlite3_column_count(statement) <= 12 else { throw StorageError.Database }
            var cells: [SqlValue] = []
            for column in 0..<sqlite3_column_count(statement) {
                switch sqlite3_column_type(statement, column) {
                case SQLITE_INTEGER: cells.append(.integer(value: sqlite3_column_int64(statement, column)))
                case SQLITE_BLOB:
                    let count = Int(sqlite3_column_bytes(statement, column)); guard count <= 4096 else { throw StorageError.Database }
                    let bytes = sqlite3_column_blob(statement, column)
                    cells.append(.bytes(value: count == 0 ? Data() : Data(bytes: bytes!, count: count)))
                case SQLITE_TEXT:
                    guard sqlite3_column_bytes(statement, column) <= 4096, let text = sqlite3_column_text(statement, column) else { throw StorageError.Database }
                    cells.append(.text(value: String(cString: text)))
                default: throw StorageError.Database
                }
            }
            rows.append(SqlRow(cells: cells))
        }
    }
}

@MainActor final class StorageVault: IdentityResetStore {
    private let folder: URL
    private let files: IdentityStorage
    private let protection: IdentityProtection
    init(folder: URL, files: IdentityStorage, protection: IdentityProtection) { self.folder = folder; self.files = files; self.protection = protection }
    static func native() throws -> StorageVault {
        let folder = try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: false).appendingPathComponent("meshchat-storage-v1", isDirectory: true)
        return StorageVault(folder: folder, files: try AppleIdentityStorage(name: "meshchat-storage-v1"), protection: AppleIdentityProtection(tag: "org.meshchat.storage.wrapping.v1"))
    }
    private func sync(_ directory: URL) throws {
        let fd = Darwin.open(directory.path, O_RDONLY); guard fd >= 0 else { throw IdentityFailure.provider }
        defer { Darwin.close(fd) }; guard fsync(fd) == 0 else { throw IdentityFailure.provider }
    }
    func access<T>(generation: Data, create: Bool, now: Int64?, work: (EncryptedStore) throws -> T) throws -> T {
        try protection.requireUnlocked()
        guard generation.count == 16 else { throw IdentityFailure.invalidInput }
        let database = folder.appendingPathComponent("history.db")
        if create {
            guard !files.hasArtifacts(), try !protection.exists() else { throw IdentityFailure.recoveryRequired }
            try files.write(.journal, Data([1])); try protection.create()
            var key = try protection.random(64); defer { key.resetBytes(in: 0..<key.count) }
            let header = Data([1]) + generation
            try files.write(.envelope, header + protection.seal(key, aad: header))
        } else {
            guard !files.exists(.journal) else { throw IdentityFailure.recoveryRequired }
            guard files.exists(.envelope), FileManager.default.fileExists(atPath: database.path) else {
                if try files.hasArtifacts() || protection.exists() { throw IdentityFailure.recoveryRequired }
                throw IdentityFailure.missing
            }
        }
        guard try protection.exists() else { throw IdentityFailure.invalidated }
        let encoded = try files.read(.envelope)
        guard (18...1024).contains(encoded.count), encoded.first == 1, Data(encoded[1..<17]) == generation else { throw IdentityFailure.invalidInput }
        var key = try protection.open(Data(encoded.dropFirst(17)), aad: Data(encoded.prefix(17)))
        defer { key.resetBytes(in: 0..<key.count) }
        guard key.count == 64 else { throw IdentityFailure.invalidInput }
        // Files inherit the protected/excluded directory; set and verify the DB
        // explicitly before releasing the creation recovery marker.
        let connection = try CipherConnection(file: database, key: key, create: create)
        let result: T
        do {
            let core = try EncryptedStore.open(db: connection, generation: generation, create: create, now: now)
            result = try work(core)
            connection.close()
        } catch { connection.close(); throw error }
        try protection.requireUnlocked()
        #if os(iOS)
        try FileManager.default.setAttributes([.protectionKey: FileProtectionType.complete], ofItemAtPath: database.path)
        // Foundation documents this dictionary value as NSString, rather than
        // the Swift RawRepresentable wrapper used when setting the attribute.
        guard try FileManager.default.attributesOfItem(atPath: database.path)[.protectionKey] as? String == FileProtectionType.complete.rawValue else { throw IdentityFailure.provider }
        #endif
        if create { try sync(folder); try files.delete(.journal) }
        return result
    }
    func clearIdentityState() throws {
        try IdentityProvider.requireResetInProgress()
        try protection.requireUnlocked()
        if try files.hasArtifacts() || protection.exists() {
            try files.write(.journal, Data([2])); try protection.delete()
            for name in ["history.db", "history.db-wal", "history.db-shm", "history.db-journal"] {
                let url = folder.appendingPathComponent(name)
                if FileManager.default.fileExists(atPath: url.path) { try FileManager.default.removeItem(at: url) }
            }
            try files.delete(.envelope); try files.delete(.journal)
            // Refuse unknown retained artifacts; never recursively remove them.
            guard try FileManager.default.contentsOfDirectory(atPath: folder.path).isEmpty else { throw IdentityFailure.recoveryRequired }
            try FileManager.default.removeItem(at: folder); try sync(folder.deletingLastPathComponent())
        }
    }
}

@MainActor final class EncryptedStorage {
    private let identity: IdentityProvider
    private let vault: StorageVault
    init(identity: IdentityProvider, vault: StorageVault) { self.identity = identity; self.vault = vault }
    /// Synchronous feature operations never retain the store/connection after return.
    func operation<T>(create: Bool = false, _ work: (EncryptedStore) throws -> T) throws -> T {
        let generation = try identity.load().identity.generation
        let now = Int64(Date().timeIntervalSince1970)
        return try vault.access(generation: generation, create: create, now: now, work: work)
    }
    func create() throws { try operation(create: true) { _ in } }
    func reopen() throws { try operation { _ in } }
    func putRecord(kind: RecordKind, key: Data, value: Data) throws { try operation { try $0.putRecord(kind: kind, key: key, value: value) } }
    func getRecord(kind: RecordKind, key: Data) throws -> Data? { try operation { try $0.getRecord(kind: kind, key: key) } }
    func deleteRecord(kind: RecordKind, key: Data) throws { try operation { try $0.deleteRecord(kind: kind, key: key) } }
    func appendUnverified(_ item: HistoryItem) throws { try operation { try $0.appendUnverified(item: item, now: Int64(Date().timeIntervalSince1970)) } }
    func acceptAuthenticated(_ item: HistoryItem, subject: Data, immutableBytes: Data) throws -> AcceptResult { try operation { try $0.acceptAuthenticated(item: item, subject: subject, immutableBytes: immutableBytes, now: Int64(Date().timeIntervalSince1970)) } }
    func history(conversation: Data, direct: Bool, limit: UInt32) throws -> [HistoryItem] { try operation { try $0.history(conversation: conversation, direct: direct, limit: limit) } }
    func deleteHistory(conversation: Data, direct: Bool) throws { try operation { try $0.deleteHistory(conversation: conversation, direct: direct) } }
    func ambiguousTarget(subject: Data, messageId: Data) throws -> Bool { try operation { try $0.ambiguousTarget(subject: subject, messageId: messageId) } }
    func prune() throws { try operation { try $0.prune(now: Int64(Date().timeIntervalSince1970)) } }
}
